keps/sig-ui: Proposal NNNN (#1)
joaquimrocha/k8s-enhancements: keps/sig-ui/NNNN-kep-template/README.md

  - [ ] (R) Minimum Two Week Window for GA e2e tests to prove flake free
- [ ] (R) Graduation criteria is in place
  - [ ] (R) [all GA Endpoints](https://github.com/kubernetes/community/pull/1806) must be hit by [Conformance Tests](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md)
- [ ] (R) Production readiness review completed
- [ ] (R) Production readiness review approved
- [ ] "Implementation History" section is up-to-date for milestone
- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes

<!--
**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
-->

[kubernetes.io]: https://kubernetes.io/
[kubernetes/enhancements]: https://git.k8s.io/enhancements
[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
[kubernetes/website]: https://git.k8s.io/website

> ## Summary

Reviewer: I find it better to leave the markdown comments here. It's simpler to review _for me_ that way.

> This proposal aims to improve the graphical user interface experience in Kubernetes by introducing a Kubernetes Desktop application and an improved in-cluster user interface, both based on the Headlamp project.

Reviewer: If you see the guidelines in the comments you deleted here, you will see it says to use a 80 chars line-length. I don't see the length in github, but it seems it is probably longer. Several other lines in the KEP seem longer too.

In fact, when trying to comment on the KEP this was an issue. Because lines are so long, I can't comment on a specific part of a line.

Also, following the guidelines, I'd also use a more direct language. Something like:

```suggestion
This KEP introduces a Kubernetes Desktop application and improved in-cluster user interface, based on the Headlamp project.
```

## Motivation

The driving motivation behind this proposal is to provide both a great out of the box experience for end users and a foundation for vendors and other projects to integrate with and build upon. Essentially, we think Headlamps' ability to be adapted and used across many scenarios allows it to serve as a central hub for the CNCF community and projects.

> ### Background

Reviewer: I'm not sure why this section is relevant here, in the motivation part. I'd just remove it.

Headlamp is currently available in two versions.
* **Headlamp** refers to the project and the default distribution of Headlamp which ships with core Kubernetes UI functionality (see the next bullet point) plus a set of plugins that the project thinks are generally useful for all users in order to provide a rich and relevant Kubernetes user experience.
* **Headlamp Base** does not include any plugins. It only includes features that the project deems universally desired in a Kubernetes UI. This is designed for builders looking to have a solid base on which to extend the user experience. They would generally create custom plugins to customize the user experience, include any existing plugins they desire and then theme the UI to fit their branding.

Each version is available as a desktop experience and an in-cluster experience. For the default build, the set of plugins differs between the in-cluster and desktop installations as there are different user profiles for each environment. For example, the plugin that allows managing plugins is not included in the in-cluster environment as the cluster admin generally determines which plugins to provide to the cluster users.

> ### Existing Dashboard project

Reviewer: I'd just remove this title and use it as motivation (removing the background section)

The Kubernetes project already has a Dashboard project. However, we think that the Headlamp project has a number of advantages that make it a better solution to build on going forward. Some of these are:
* Has a robust plugin system for customization and integrations
* Supports both desktop and in-cluster mode
* Is under very active development and has dedicated resources assigned to it in addition to the active external contributors
* Is built on a very common technology stack (Typescript, React, Material UI, Vite, Go)
* Has more advanced features (multi-cluster, graph view, powerful plugin system, 3rd party plugins)
* Addresses many more use cases (see below)

### Goals

With this proposal we aim to provide the Kubernetes project with the following.
* A tool to more easily onboard new and beginner-level users
* A UI that serves as a base for vendors to build upon
* A more modern UI developer environment
* A more feature-rich UI
* A desktop user experience
* An integration point for other CNCF and external projects, especially those without a UI focus

### Non-Goals

> This proposal does not propose to force people who are using the current Dashboard to move to Headlamp.

Reviewer: Use this section too. Try to think "defensive" on things that might be asked and you don't want to tackle as part of the initial KEP. Think of people asking for this later on, and you can point out "that was stated as a non-goal" of the KEP that is approved.

 I don't know about the KEP and what is reasonable to do or not, but I can imagine a non-goal might be to not support automatic migration of the current k8s dashboard, for example.

Try to really use this section to narrow the scope of the KEP, to clarify some of the goals even (like "we say X is a goal, but Y is not part of the goal")

The current repo kubernetes/dashboard would remain but include a note about the existence of the other repo. This would allow for users to compare the projects and make an informed decision about which works best for their use case. Over time, we can measure adoption trends and make decisions based on that data.

The Headlamp branding would be phased out.

### User Stories (Optional)

One of Headlamp's strengths is that it was designed to both deliver a great end-user experience and be a platform for others to build upon and integrate with. This is demonstrated by the wide range of user stories it enables  and one of the reasons we believe it can become a hub for the community.

_**Note:** Using “Headlamp” in the following examples to avoid confusion._

#### Story 1

As an end user I want to get an overview of my cluster. I start the Headlamp Desktop app which picks up my clusters from my kubeconfig and I’m now able to explore my clusters.

#### Story 2

> As a new Kubernetes user I want to learn more about Kubernetes. I use the Headlamp Desktop app to guide me in installing a local kubernetes cluster using minikube. Once my local cluster is installed I’m presented with a tutorial plugin which guides me through the various aspects of the UI and Kubernetes concepts.

Reviewer: heh, it's hard to comment on the right part because it is a very long line.

I'd use this section as more high level. The high-level details are what matter here.

On one side, if you are planning to guide the installation of minikube with a graphical UI and in a few clicks it will just work, I'd highlight that here. That is great, it is clear how simple it is expected to be and how that is an improvement over today's experience. If you plan to say: "this is the link to the the minikube install instructions, install it and come back", I'd say it too.

These are the classic things that might be very clear to you in your head, but maybe not for others.

> As a new Kubernetes user I want to learn more about Kubernetes. I use the Headlamp Desktop app to guide me in installing a local kubernetes cluster using minikube. Once my local cluster is installed I’m presented with a tutorial plugin which guides me through the various aspects of the UI and Kubernetes concepts.

Reviewer: I'd not say if the tutorial is a plugin or not here. For the user-story, the implementation details are not really relevant. If you want to highlight that, in the proposal section you can say the tutorial will be a plugin and so.

_**Note:** We have the minikube installer plugin in POC but the tutorial plugin is not something we currently have._

#### Story 3

> As a cluster admin I was to provide a user interface for my users. I choose which plugins I want to include for my users and deploy Headlamp to my cluster. I have peace of mind knowing that Headlamp respects the RBAC rules that each user has.

Reviewer: If I'm understanding this properly, it took me a while. I had to read this, the later section on auth, and re-read this user story to understand it. I think the user story should stand on itself instead, though.

It seems to me, IIUC, that this usre story tries to mention these things:
 * I want to provide a UI for users and select what headlamp features can be used (or similar, what you want to do with plugins)
 * I want to provide a UI that is not a security nightmare that needs to run with a super-privileged service account
 * I want to provide a UI that only allows users to do what their RBAC permissions allow.

Am I understanding it right?

Maybe splitting it in more user stories would be simpler.

And maybe that would also allow to remove the "reference text" later explaining how auth work? It's not easy to get it if you need to read this, read the auth text below and then re-read it to realize the subtle things this implies.

> As a cluster admin I was to provide a user interface for my users. I choose which plugins I want to include for my users and deploy Headlamp to my cluster. I have peace of mind knowing that Headlamp respects the RBAC rules that each user has.

Reviewer: Does the current k8s dashboard need an all powerful service account to run? In that case, I'd highlight headlamp doesn't even more.

#### Story 5

As a CNCF or external  project, I want to provide my users with a graphical experience but I don’t want to provide a full blown UI. I create a plugin that can be as simple as adding a column to a table (ex. OpenCost plugin) or as complex as creating a dedicated view with a sidebar entry and integrations across the various resource types (ex. Flux plugin).
Example: See the OpenCost and Flux plugins

#### Story 6

As a CNCF or external project I have a dedicated UI but that UI is not providing a full Kubernetes context. To provide my users with a Kubernetes context I create a link from my project to the Headlamp instance running in the cluster and create a plugin for Headlamp that provides the user in context views or links to my projects resources. This allows the user to easily navigate between the projects. (Ex. Backstage plugin)

The current Kubernetes Dashboard only supports part of Story 3.

> ### Technical Overview

Reviewer: FYI: You can add your own sections, but you will see tests don't pass. You will need to run a script from the top of the repo to update the TOC at the beginning based on the sections.

> ### Technical Overview

Reviewer: I'm not sure why you are adding this section. Is this something that was asked by the SIG?

If it is not asked by the SIG, I'm unsure if I'd include it. Or at least make it way shorter. See user story 3 for ideas that, IMHO, if said there can remove this section completely

If you want to keep it, though, things like:
```
In the case of the user’s kubeconfig, the contexts’ related user information is also kept in the backend.
```

seems to go in quite some detail, sharing some design decisions, but if you don't want people to question those (because headlamp it is not in a design state), I'd not go into such detail. If you are going in detail for some reason (maybe it is stated and I missed it!), like to explain the security benefits, I'd just say that ("headlamp doesn't need a privileged service account to run, it relies on the k8s user credentials...")

> Headlamp chose the most common technologies used in the community and wider space in order to lower the barriers to contributing as much as possible.

Reviewer: Sentences like "the most common technologies" are quite open for debate, especially as trends can vary quickly. I'd use more neutral vocabulary, trying to avoid subjective opinions stated as facts.

This probably applies to more parts of the KEP.
